src/Distribution/Audit.hs

```diff
@@ -194,6 +208,80 @@ osvHandler mkHandle mp =
               ]
         ]
 
+sarifHandler
+  :: MonadUnliftIO m
+  => Codensity IO Handle
+  -> ProjectBaseContext
+  -> [(PackageName, ElaboratedPackageInfoAdvised)]
+  -> m ()
+sarifHandler mkHandle projectBaseContext packageAdvisories = do
+  let projectRoot = distProjectRootDirectory $ distDirLayout projectBaseContext -- TODO(blackheaven): resolve repository root
+  ghcVersion <- liftIO $ T.pack <$> readProcess "ghc" ["--version"] ""
+  let advisories =
+        M.elems $
+          M.fromListWith (\(advisory, pkgsInfo) (_, pkgsInfo') -> (advisory, pkgsInfo <> pkgsInfo')) $
+            flip concatMap packageAdvisories $ \(pkgName, pkgInfo) ->
+              flip map (runIdentity pkgInfo.packageAdvisories) $ \(advisory, fixedAt) ->
+                (advisory.advisoryId, (advisory, [(pkgName, fixedAt)]))
+      run =
+        MkRun
+          { runTool =
+              let tool name version =
+                    defaultToolComponent
+                      { toolComponentName = Just name,
+                        toolComponentVersion = Just version
+                      }
+               in MkTool
+                    { toolExtensions =
+                        [ tool "hsec-tools" VERSION_hsec_tools,
+                          tool "ghc" ghcVersion
+                        ],
+                      toolDriver = tool "cabal-audit" VERSION_cabal_audit
+                    },
+            runResults =
+              flip map advisories $ \(advisory, concernedInfo) ->
+                MkResult
+                  { resultRuleId = T.pack $ printHsecId advisory.advisoryId,
+                    resultMessage =
+                      defaultMultiformatMessageString $
+                        let hsecId = T.pack (printHsecId advisory.advisoryId)
+                         in T.intercalate "\n" $
+                              concat @[]
+                                [ [hsecId <> " \"" <> advisory.advisorySummary <> "\""],
+                                  ["published: " <> T.pack (show advisory.advisoryPublished)],
+                                  ["https://haskell.github.io/security-advisories/advisory/" <> hsecId],
+                                  ["Concerned package:"],
+                                  concernedInfo <&> \(pkgName, fixedAt) ->
+                                    "* "
+                                      <> T.pack (unPackageName pkgName)
+                                      <> ": "
+                                      <> case fixedAt of
+                                        Nothing -> "No fix version available"
+                                        Just fv -> "Fix available since version " <> prettyVersion fv,
+                                  [T.intercalate ", " (coerce advisory.advisoryKeywords)]
```

Comment:
can we maybe reuse the messages used for the human readable output to make it a bit more streamlined? doesn't need to be, I can do this in a follow up :)

Reply:
it was a little involved, but it's done